kafka/consumer/fetcher.py

```diff
@@ -935,21 +877,55 @@ def __init__(self, fetch_offset, tp, messages):
                 if msg.offset == fetch_offset:
                     self.message_idx = i
 
+        # For truthiness evaluation we need to define __len__ or __nonzero__
+        def __len__(self):
+            if self.messages is None or self.message_idx >= len(self.messages):
+                return 0
+            return len(self.messages) - self.message_idx
+
         def discard(self):
             self.messages = None
 
-        def take(self, n):
-            if not self.has_more():
+        def take(self, n=None):
+            if not len(self):
                 return []
+            if n is None or n > len(self):
+                n = len(self)
             next_idx = self.message_idx + n
             res = self.messages[self.message_idx:next_idx]
             self.message_idx = next_idx
-            if self.has_more():
+            if len(self) > 0:
                 self.fetch_offset = self.messages[self.message_idx].offset
             return res
 
-        def has_more(self):
-            return self.messages and self.message_idx < len(self.messages)
+
+class FetchResponseMetricAggregator(object):
```

Comment:
Will this aggregator behave properly in case CompletedFetch is not parsed? Say we parse 2 of 3 partitions and the user uses `seek()` which will cause the CompletedFetch to be discarded.